extensions/trace-propagators/src/main/java/io/opentelemetry/extension/trace/propagation/B3PropagatorExtractorMultipleHeaders.java

```diff
@@ -28,33 +28,37 @@
       Context context, C carrier, TextMapPropagator.Getter<C> getter) {
     Objects.requireNonNull(carrier, "carrier");
     Objects.requireNonNull(getter, "getter");
-    SpanContext spanContext = getSpanContextFromMultipleHeaders(carrier, getter);
-    if (!spanContext.isValid()) {
-      return Optional.empty();
-    }
-
-    return Optional.of(context.with(Span.wrap(spanContext)));
+    return extractSpanContextFromMultipleHeaders(context, carrier, getter);
   }
 
-  private static <C> SpanContext getSpanContextFromMultipleHeaders(
-      C carrier, TextMapPropagator.Getter<C> getter) {
+  private static <C> Optional<Context> extractSpanContextFromMultipleHeaders(
+      Context context, C carrier, TextMapPropagator.Getter<C> getter) {
     String traceId = getter.get(carrier, TRACE_ID_HEADER);
     if (StringUtils.isNullOrEmpty(traceId)) {
-      return SpanContext.getInvalid();
+      return Optional.empty();
     }
     if (!Common.isTraceIdValid(traceId)) {
       logger.fine(
           "Invalid TraceId in B3 header: " + traceId + "'. Returning INVALID span context.");
-      return SpanContext.getInvalid();
+      return Optional.empty();
     }
 
     String spanId = getter.get(carrier, SPAN_ID_HEADER);
     if (!Common.isSpanIdValid(spanId)) {
       logger.fine("Invalid SpanId in B3 header: " + spanId + "'. Returning INVALID span context.");
-      return SpanContext.getInvalid();
+      return Optional.empty();
+    }
+
+    // if debug flag is set, then set sampled flag, and also store the B3 debug flag in the context
+    // for onward use by B3 injector
+    if (Common.TRUE_INT.equals(getter.get(carrier, DEBUG_HEADER))) {
```

Comment:
Shouldn't this be comparing to B3Propagator.DEBUG_SAMPLED, since we're extracting it from the incoming request?

Reply:
er wait. that's `"d"` where as the format specifies a "1" for the debug flag. What is the "d" for?

Reply:
I got it. "d" is used for single-header and "1" is used for multi-header. I recommend not using the `Common` constant here, but putting a separate constant in the B3Propagator class for B3 multi-header in particular.

Reply:
Sure, what I've done is add a couple of constants in B3Propagator which should hopefully make this clearer
```
  static final String MULTI_HEADER_DEBUG = "1";
  static final String SINGLE_HEADER_DEBUG = "d";
```